### PR TITLE
[te] Ban uint8 tensors from fusion groups

### PR DESCRIPTION
**Pull request by bertmaher**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49271 [te] Fix bugs with shift operators
* **#49247 [te] Ban uint8 tensors from fusion groups**

uint8's expose all kind of corner cases in type promotion.  As an example, consider:
```
>>> torch.tensor([1], dtype=torch.uint8).lt(-1)
tensor([True])
>>> torch.tensor([1], dtype=torch.uint8).lt(torch.tensor(-1))
tensor([True])
>>> torch.tensor([1], dtype=torch.uint8).lt(torch.tensor([-1]))
tensor([False])
```
the difference is how promotions involving scalars (or 0-dim tensors, which are treated like scalars) are prioritized compared to tensor dtypes.
Per @eellison, the order is something like:
1. Tensor FP types
2. Scalar FP types
3. Tensor Int types
4. Scalar Int types

The logic for this is here: https://github.com/pytorch/pytorch/blob/c73e97033a3aef97a5685588ea014d54a5cc11cc/aten/src/ATen/native/TypeProperties.cpp#L93

AFAICT the effects are mainly visible for the unsigned byte type (the only unsigned type, besides bool) since the others degrade more or less gracefully.

It's hard to re-use this logic as is in TensorIterator/TypeProperties, and it's complicated enough that it's not worth re-implementing in TE unless there's evidence that it matters for real models.

Differential Revision: [D25489035](https://our.internmc.facebook.com/intern/diff/D25489035/)

### CODE REVIEW

File: test/test_jit_fuser_te.py [hunk 1]
```diff
@@ -490,7 +490,8 @@ def apply(fn):
             try:
                 t = torch.jit.trace(fn, (x, y, z))
                 self.assertEqual(ref, t(x, y, z))
-                self.assertAllFused(t.graph_for(x, y, z))
+                if dtype != torch.uint8:
```
Comment (eellison) [inline test/test_jit_fuser_te.py:493]: can we move the dtypes to `self.dtypes` and exclude `torch.uint8` ?

**Review by eellison — APPROVED** [review]
LGTM! i think we should be able to accurately model pytorch dtype promotion eventually, without it leaking out of kernel.cpp

File: test/test_jit_fuser_te.py (hunk 1 above)
Comment (bertmaher) [reply test/test_jit_fuser_te.py:493]: Oh yeah that's a better idea than this.  I had kind of wanted to test a few uint cases just to make sure nothing was broken but no reason to exhaustively go through them here.
